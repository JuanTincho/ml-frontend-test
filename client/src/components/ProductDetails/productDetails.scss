@import "../../styles/colors.scss";

.product-container {
  background: #ffffff;
  border-radius: 4px;
  color: $color-gray-dark;
  display: grid;
  grid-template-columns: repeat(10, 1fr);
  grid-row-gap: 40px;

  img {
    border-radius: 4px;
    grid-column: 1 / 8;
    margin: 30px 40px;
    text-align: initial;
    width: 680px;
  }

  .details {
    margin-top: 32px;
    display: grid;
    grid-column: 8 / -1;
    grid-template-columns: repeat(3, 1fr);
    grid-template-rows: repeat(auto-fit, 200px);
    grid-row-gap: 32px;

    &_column {
      &-1 {
        grid-column: 1 / 3;

        .condition {
          color: $color-gray-medium;
          font-size: 14px;
          margin-bottom: 16px;
        }

        .title {
          font-size: 24px;
        }

        .price {
          font-size: 46px;
          margin-top: 32px;

          // margin-bottom: 32px;
       
        }
      }

      &-2 {
        grid-column: 1 / -1;
        margin-right: 32px;

        button {
          background: $color-secondary;
          border: none;
          border-radius: 4px;
          color: #ffffff;
          cursor: pointer;
          font-size: 16px;
          padding: 15px 32px;
          width: 100%;
        }
      }
    }
  }

  .description {
    grid-column: 1 / 8;
    margin-left: 32px;
    h1 {
      font-size: 28px;
    }

    p {
      color: $color-gray-medium;
      font-size: 16px;
      margin: 32px 0;
      white-space: pre-line;
    }
  }
}
