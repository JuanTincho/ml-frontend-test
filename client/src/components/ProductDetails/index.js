import React, { Component } from 'react';
import PropTypes from 'prop-types';
import axios from 'axios';
import Categories from '../Categories';
import './productDetails.scss';
// import getParams from '../../utils/getParams';

class ProductDetails extends Component {
  constructor(props) {
    super(props);

    this.state = {
      item: null,
    };
  }

  componentDidMount() {
    const {
      match: {
        params: { id },
      },
    } = this.props;

    this.fetchData(id);
  }

  componentDidUpdate(prevProps) {
    const {
      match: {
        params: { id },
      },
    } = this.props;

    const {
      match: {
        params: { id: prevId },
      },
    } = prevProps;

    if (id !== prevId) {
      this.fetchData(id);
    }
  }

  fetchData = (itemId) => {
    axios.get(`/api/items/${itemId}`).then((response) => {
      const { item } = response.data;
      this.setState({ item });
    });
  };

  render() {
    const { item } = this.state;
    return (
      <>
        {item ? (
          <>
            <Categories categories={item.categories} />
            <div className="product-container">
              <img src={item.picture} alt={item.title} />
              <div className="details">
                <div className="details_column-1">
                  <p className="condition">
                    {`${item.condition === 'new' ? 'Nuevo' : 'Usado'} - 
                  ${item.sold_quantity} vendidos`}
                  </p>
                  <h1 className="title">{item.title}</h1>
                  <p className="price">
                    {new Intl.NumberFormat([], {
                      style: 'currency',
                      currency: item.price.currency,
                    }).format(item.price.amount)}
                  </p>
                </div>
                <div className="details_column-2">
                  <button type="button"> Comprar </button>
                </div>
              </div>
              <div className="description">
                <h1>Descripci√≥n del producto</h1>
                <p>{item.description}</p>
              </div>
            </div>
          </>
        ) : null}
      </>
    );
  }
}

ProductDetails.propTypes = {
  // eslint-disable-next-line react/forbid-prop-types
  match: PropTypes.object.isRequired,
};
export default ProductDetails;
