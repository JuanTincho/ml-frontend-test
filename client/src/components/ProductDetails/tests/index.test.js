import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import mockedAxios from 'axios';
import ProductDetails from '../index';

function flushPromises() {
  return new Promise(resolve => setImmediate(resolve));
}

describe('ProductDetails', () => {
  const mockedId = 'mockedId';
  const mockedMatch = { params: { id: mockedId } };
  let component;

  beforeEach(() => {
    component = shallow(<ProductDetails match={mockedMatch} />);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  describe('render', () => {
    it('should match snapshotwith no item', () => {
      const tree = renderer.create(component).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('should match snapshotwith with an item ', async () => {
      mockedAxios.get.mockResolvedValue({
        data: {
          item: {
            title: 'mockedTitle',
            picture: 'mockedPic',
            price: {
              currency: 'ARS',
              amount: 500,
            },
            condition: 'new',
            freeShipping: true,
            description: 'Descripcion',
            sold_quantity: 12,
          },
        },
      });

      component = shallow(<ProductDetails match={mockedMatch} />);
      await flushPromises();
      const tree = renderer.create(component).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  it('should call to the service', () => {
    const expectedCall = `/api/items/${mockedId}`;
    expect(mockedAxios.get).toHaveBeenCalledTimes(1);
    expect(mockedAxios.get).toHaveBeenCalledWith(expectedCall);
  });
});
