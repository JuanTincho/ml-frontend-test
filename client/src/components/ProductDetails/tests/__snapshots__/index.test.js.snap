// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductDetails render should match snapshotwith no item 1`] = `null`;

exports[`ProductDetails render should match snapshotwith with an item  1`] = `
Array [
  <div
    className="categories"
  />,
  <div
    className="product-container"
  >
    <img
      alt="mockedTitle"
      src="mockedPic"
    />
    <div
      className="details"
    >
      <div
        className="details_column-1"
      >
        <p
          className="condition"
        >
          Nuevo - 
                  12 vendidos
        </p>
        <h1
          className="title"
        >
          mockedTitle
        </h1>
        <p
          className="price"
        >
          ARS 500.00
        </p>
      </div>
      <div
        className="details_column-2"
      >
        <button
          type="button"
        >
           Comprar 
        </button>
      </div>
    </div>
    <div
      className="description"
    >
      <h1>
        Descripción del producto
      </h1>
      <p>
        Descripcion
      </p>
    </div>
  </div>,
]
`;
